src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.ConnectionPools.cs

```diff
@@ -0,0 +1,347 @@
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Execution;
+using Microsoft.CodeAnalysis.Extensions;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.ServiceHub.Client;
+using Roslyn.Utilities;
+using StreamJsonRpc;
+
+namespace Microsoft.VisualStudio.LanguageServices.Remote
+{
+    internal sealed partial class ServiceHubRemoteHostClient : RemoteHostClient
+    {
+        private class ConnectionPools
+        {
+            private readonly HubClient _hubClient;
+            private readonly HostGroup _hostGroup;
+            private readonly TimeSpan _timeout;
+
+            private readonly ReaderWriterLockSlim _shutdownLock;
+            private readonly ReferenceCountedDisposable<RemotableDataJsonRpc> _remotableDataRpc;
+
+            private readonly int _maxPoolConnections;
+            private readonly ConcurrentDictionary<string, ConcurrentQueue<JsonRpcConnection>> _pools;
+
+            // indicate whether pool should be used.
+            // it is mutable since it will set to false when this pool got shutdown
+            private volatile bool _usePool;
+
+            public ConnectionPools(
+                HubClient hubClient,
+                HostGroup hostGroup,
+                bool usePool,
+                int maxPoolConnection,
+                TimeSpan timeout,
+                ReferenceCountedDisposable<RemotableDataJsonRpc> remotableDataRpc)
+            {
+                _hubClient = hubClient;
+                _hostGroup = hostGroup;
+                _timeout = timeout;
+                _remotableDataRpc = remotableDataRpc;
+
+                _maxPoolConnections = maxPoolConnection;
+                _pools = new ConcurrentDictionary<string, ConcurrentQueue<JsonRpcConnection>>(concurrencyLevel: 4, capacity: 4);
+
+                _usePool = usePool;
+                _shutdownLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
+            }
+
+            public Task<Connection> TryCreateConnectionAsync(string serviceName, object callbackTarget, CancellationToken cancellationToken)
+            {
+                // pool is turned off either by option or pool has shutdown.
+                if (!_usePool)
+                {
+                    return TryCreateNewConnectionAsync(serviceName, callbackTarget, cancellationToken);
+                }
+
+                // when callbackTarget is given, we can't share/pool connection since callbackTarget attaches a state to connection.
+                // so connection is only valid for that specific callbackTarget. it is up to the caller to keep connection open
+                // if he wants to reuse same connection
+                if (callbackTarget != null)
+                {
+                    return TryCreateNewConnectionAsync(serviceName, callbackTarget, cancellationToken);
+                }
+
+                return TryGetConnectionFromPoolAsync(serviceName, callbackTarget, cancellationToken);
```

Comment:
callbackTarget is always null here?  so why is it necessary to pass that value along?

Reply:
yes. forgot to remove it.